### PR TITLE
Feature/personalization

### PR DESCRIPTION
**Pull request by yuvalron**
Added the ability to mark a question as solved.
Once the user marks a question and then returns to the same question page, then, of course, the question will already be marked. 

All marked questions appear on the profile page.

I also improved the course page. Now instead of a list of questions and exams, there is now a table.
The table has a sorting feature for every column. I also moved the forum to the bottom of the page since it is less important than the available questions and exams.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "@angular/platform-browser-dynamic": "^7.2.13",
     "@angular/router": "^7.2.13",
     "angular-font-awesome": "^3.1.2",
+    "angular-star-rating": "^4.0.0-beta.3",
```
Comment (ron4548) [inline package.json]: You don't use this package. Remove it.

File: src/app/core/db.service.ts [hunk 2]
```diff
@@ -134,6 +135,44 @@ export class DbService {
       }));
   }
 
+  // getSolvedQuestion(uId: string, qId: string): Observable<SolvedQuestionId[]> {
```
Comment (ron4548) [inline src/app/core/db.service.ts]: Delete this function or uncomment it. Do not leave code in a comment.

File: src/app/ui/question/question.component.ts [hunk 3]
```diff
@@ -16,22 +19,54 @@ import {Course} from '../../entities/course';
...
+  constructor(private route: ActivatedRoute, private db: DbService, private auth: AuthService, private snackBar: MatSnackBar) {
+    this.qId = this.route.snapshot.paramMap.get('id');
+    this.topics$ = this.db.getTopicsForQuestion(this.qId);
+    this.question$ = this.db.getQuestion(this.qId);
+    this.solutions$ = this.db.getSolutions(this.qId);
+    this.isAdmin$ = this.db.getQuestion(this.qId).pipe(flatMap(q => this.auth.isAdminForCourse(q.course)));
+    this.course$ = this.db.getQuestion(this.qId).pipe(flatMap(q => this.db.getCourse(q.course)));
+    this.userId = this.auth.currentUserId;
+    this.solvedQuestions$ = this.db.getSolvedQuestions(this.userId);
+    this.solvedQuestions$.subscribe(value => {
```
Comment (ron4548) [inline src/app/ui/question/question.component.ts]: This is bad. You query for all the questions the user solved just to check if he solved a single question? Check it directly without going over all the questions..

File: src/app/users/user-profile/user-profile.component.html [hunk 4]
```diff
@@ -20,3 +19,42 @@ <h1 class="display-4">
...
+  </tr>
+  <tr class="table-light text-center" >
+    <!--<th scope="col"></th>-->
+    <th scope="col" mat-sort-header="course">Course</th>
+    <th scope="col" mat-sort-header="year">Year</th>
+    <th scope="col" mat-sort-header="semester">Semester</th>
+    <th scope="col" mat-sort-header="moed">Moed</th>
+    <th scope="col" mat-sort-header="number">Question No.</th>
+    <th scope="col" mat-sort-header="difficulty">Difficulty</th>
+    <!--<th scope="col">Remove?</th>-->
```
Comment (ron4548) [inline src/app/users/user-profile/user-profile.component.html]: ??

File: src/app/users/user-profile/user-profile.component.html [hunk 5]
```diff
@@ -20,3 +19,42 @@ <h1 class="display-4">
...
+<table matSort (matSortChange)="sortQuestions($event)" class="table table-hover" *ngIf="questions" >
+  <thead>
+  <tr class="table-success">
+    <th colspan="9">
+      Questions you solved:
+      <span class="badge badge-success">{{questions.length}}</span>
+    </th>
+  </tr>
+  <tr class="table-light text-center" >
+    <!--<th scope="col"></th>-->
```
Comment (ron4548) [inline src/app/users/user-profile/user-profile.component.html]: ??

File: src/app/users/user-profile/user-profile.component.html [hunk 6]
```diff
@@ -20,3 +19,42 @@ <h1 class="display-4">
...
+    <th scope="col" mat-sort-header="semester">Semester</th>
+    <th scope="col" mat-sort-header="moed">Moed</th>
+    <th scope="col" mat-sort-header="number">Question No.</th>
+    <th scope="col" mat-sort-header="difficulty">Difficulty</th>
+    <!--<th scope="col">Remove?</th>-->
+  </tr>
+  </thead>
+  <tbody>
+  <tr *ngFor="let q of questions">
+    <!--<td class="clickable" routerLink="/question/{{ q.id }}">-->
```
Comment (ron4548) [inline src/app/users/user-profile/user-profile.component.html]: ??

File: src/app/users/user-profile/user-profile.component.html [hunk 7]
```diff
@@ -20,3 +19,42 @@ <h1 class="display-4">
...
+      <!--<fa name="external-link-square" style="color: forestgreen"></fa>-->
+    <!--</td>-->
+    <td>{{ q.course }}</td>
+    <td>{{ q.year }}</td>
+    <td>{{ q.semester | titlecase}}</td>
+    <td>{{ q.moed | titlecase}}</td>
+    <td>{{ q.number }}</td>
+    <td></td>
+    <!--<td>-->
+      <!--<button class="btn btn-danger btn-circle mr-2" type="button" (click)="remove()">-->
```
Comment (ron4548) [inline src/app/users/user-profile/user-profile.component.html]: ??

File: src/app/ui/question/question.component.ts (hunk 3 above)
Comment (ron4548) [reply src/app/ui/question/question.component.ts]: And make it observable, do not subscribe in the code but use async pipe.

Comment (yuvalron) [reply src/app/ui/question/question.component.ts]: > And make it observable, do not subscribe in the code but use async pipe.

The async pipe is used in the HTML file, right?

Comment (yuvalron) [reply src/app/ui/question/question.component.ts]: What it the problem with subscribing?

**Pull request merged**